saw-core/src/Verifier/SAW/Term/Functor.hs

```diff
@@ -386,15 +386,21 @@ data TermF e
   deriving (Eq, Ord, Show, Functor, Foldable, Traversable, Generic)
 
 -- See the commentary on 'Hashable Term' for a note on uniqueness.
-instance Hashable e => Hashable (TermF e) -- automatically derived.
--- NB: we may someday wish to customize this instance, for a couple reasons.
+instance Hashable e => Hashable (TermF e) where
+  hashWithSalt salt (FTermF ftf) = salt `hashWithSalt` ftf
+  hashWithSalt salt (App _ t u) = salt `hashWithSalt` t `hashWithSalt` u
+  hashWithSalt salt (Lambda _ t u) = salt `hashWithSalt` t `hashWithSalt` u
+  hashWithSalt salt (Pi _ t u) = salt `hashWithSalt` t `hashWithSalt` u
+  hashWithSalt salt (LocalVar i) = salt `hashWithSalt` i
+  hashWithSalt salt (Constant ec _) = salt `hashWithSalt` ec
```

Comment:
This deserves a comment explaining why we write this instance in such a peculiar why, including a reference to the `alphaEquiv` function (on which this is based).